First sprint

ZernovNikolay opened the pull request.

ZernovNikolay committed:
Version 1.0 of first sprint

Nadezhda94 commented on include/crypto_guard_ctx.h:
```diff
@@ -16,13 +17,16 @@ class CryptoGuardCtx {
+    std::unique_ptr<Impl> pImpl_;
+
+private:
+    enum CifherMode { DECRYPT, ENCRYPT };
```
Это тоже детали реализации, в интерфейсе API никак не используется, их лучше перенести в Impl.

Nadezhda94 commented on src/cmd_options.cpp:
```diff
@@ -1,11 +1,61 @@
+    desc_.add_options()("help,h", "Show help message")("command,c", po::value<std::string>()->required(),
+                                                       "Command to execute: encrypt, decrypt or checksum")(
+        "input,i", po::value<std::string>()->required(), "Input file path")(
+        "output,o", po::value<std::string>()->required(), "Output file path")("password,p", po::value<std::string>(),
```
output не всегда используется, например, для команды checksum

Nadezhda94 commented on src/cmd_options.cpp:
```diff
@@ -1,11 +1,61 @@
+namespace po = boost::program_options;
+
+ProgramOptions::ProgramOptions() : desc_("Allowed options") {
+    desc_.add_options()("help,h", "Show help message")("command,c", po::value<std::string>()->required(),
```
👍 required для command

Nadezhda94 commented on src/cmd_options.cpp:
```diff
@@ -1,11 +1,61 @@
+    } catch (const po::error &e) {
+        std::cerr << "Error: " << e.what() << "\n";
+        std::cerr << desc_ << "\n";
+        exit(1);
```
exit() тоже завершит всю программу. Предложила бы не ловить исключение здесь, пусть оно летит дальше в main, а там его поймаем и выведем сообщение об ошибке.

Nadezhda94 commented on src/cmd_options.cpp:
```diff
@@ -1,11 +1,61 @@
+            std::ostringstream oss;
+            oss << desc_;
+            std::print(stdout, "{}", oss.str());
+            exit(0);
```
Вызов exit() завершит выполнение всей программы. Тут лучше просто сделать return, с точки зрения класса options help валидная команда.

Nadezhda94 commented on src/cmd_options.cpp:
```diff
@@ -1,11 +1,61 @@
+        outputFile_ = vm["output"].as<std::string>();
+
+        if (command_ == COMMAND_TYPE::ENCRYPT || command_ == COMMAND_TYPE::DECRYPT) {
+            if (!vm.count("password")) {
```
Тут тоже лучше использовать find, потому что `vm.count("password")` и `vm["password"]` выполнят два поиска по контейнеру. После замены на find получится один поиск.

Nadezhda94 commented on src/cmd_options.cpp:
```diff
@@ -1,11 +1,61 @@
+
+        // Set other parameters
+        inputFile_ = vm["input"].as<std::string>();
+        outputFile_ = vm["output"].as<std::string>();
```
output не нужен для checksum

Nadezhda94 commented on src/crypto_guard_ctx.cpp:
```diff
@@ -1,4 +1,237 @@
+public:
+    StreamGuard(std::iostream &stream, std::string_view name) : stream_(stream) {
+        if (!stream_.good())
+            throw std::runtime_error(std::string(name) + " stream is in bad state");
```
Можно воспользоваться std::format
`std::runtime_error( std::format("{} stream is in bad state"), name ) )`

Nadezhda94 commented on src/crypto_guard_ctx.cpp:
```diff
@@ -1,4 +1,237 @@
+
+    ~Impl() { EVP_cleanup(); }
+
+    struct EVP_Cipher_Deleter {
```
Лучше перенести в private

Nadezhda94 commented on src/crypto_guard_ctx.cpp:
```diff
@@ -1,4 +1,237 @@
+    return oss.str();
+}
+
+void CryptoGuardCtx::Impl::ProcessCifher(std::unique_ptr<EVP_CIPHER_CTX, EVP_Cipher_Deleter> ctx, std::iostream &input,
```
unique_ptr корректно передается и принимается. 👍 
Но посоветовала бы тут лучше принимать EVP_CIPHER_CTX по ссылке. Владение указателем можно внутрь функции не отдавать.

Nadezhda94 commented on src/crypto_guard_ctx.cpp:
```diff
@@ -1,4 +1,237 @@
+void CryptoGuardCtx::Impl::ProcessCifher(std::unique_ptr<EVP_CIPHER_CTX, EVP_Cipher_Deleter> ctx, std::iostream &input,
+                                         std::iostream &output) {
+
+    size_t buf_size = 1024;
```
const size_t

Nadezhda94 commented on src/crypto_guard_ctx.cpp:
```diff
@@ -1,4 +1,237 @@
+};
+
+std::string get_openssl_error() {
+    std::vector<char> buf(256);
```
Тут даже лучше подойдет std::array

Nadezhda94 commented on src/crypto_guard_ctx.cpp:
```diff
@@ -1,4 +1,237 @@
+        if (inStream.eof())
+            break;
+
+        if (!inStream.good())
```
Лучше проверить после того как прочитали, если чтение завершилось с ошибкой, то уже можно не продолжать, т.к. данные могут быть испорчены.

Nadezhda94 commented on src/crypto_guard_ctx.cpp:
```diff
@@ -1,4 +1,237 @@
+        if (input.eof())
+            break;
+
+        if (!input.good()) {
```
Лучше проверить сразу после чтения и до шифрования, если read завершилось с ошибкой, то можем получить испорченные данные.

Nadezhda94 commented on src/crypto_guard_ctx.cpp:
```diff
@@ -1,4 +1,237 @@
+        StreamGuard guard_out(outStream, "Output");
+    }
+
+    ERR_clear_error();
```
👍

Nadezhda94 commented on src/crypto_guard_ctx.cpp:
```diff
@@ -1,4 +1,237 @@
+                                       CryptoGuardCtx::CifherMode mode) {
+
+    {
+        StreamGuard guard_in(inStream, "Input");
```
Здорово, что проверка потоков идет до шифрования!
В таком формате можно просто оставить проверку `!stream_.good()` без  обертки. Возможно она была нужна для отладки.

Nadezhda94 commented on src/crypto_guard_ctx.cpp:
```diff
@@ -1,4 +1,237 @@
+        OpenSSL_add_all_digests();
+    }
+
+    ~Impl() { EVP_cleanup(); }
```
Правило 5. Перемещающие операции можно сделать default, копирующие delete

Nadezhda94 commented on src/main.cpp:
```diff
@@ -81,6 +81,7 @@ int main(int argc, char *argv[]) {
         //
```
AesCipherParams и CreateChiperParamsFromPassword нужно отсюда удалить.

Nadezhda94 replied:
И пример использования openssl.

Nadezhda94 commented on src/main.cpp:
```diff
@@ -81,6 +81,7 @@ int main(int argc, char *argv[]) {
+        options.Parse(argc, argv);
 
         CryptoGuard::CryptoGuardCtx cryptoCtx;
 
```
Не хватает обработки команд в switch ниже.

Nadezhda94 commented on tests/cmd_options_test.cpp:
```diff
@@ -0,0 +1,141 @@
+}
+
+TEST(ProgramOptions, HelpOption) {
+    auto [argc, argv] = create_argv({"program", "--help"});
```
Лучше не использовать сырые указатели и new, delete в явном виде. 
Если EXPECT_... не выполнится, то будет утечка памяти.
Тут может пригодиться std::array<const char *, 2>
С помощью метода data() можно будет получить указатель на внутренний буфер

Nadezhda94 commented on tests/crypo_guard_ctx_test.cpp:
```diff
@@ -1,3 +1,242 @@
+    // Проверка совпадения начальной и конечной контрольной суммы
+    std::stringstream input_second(plaintext_);
+    decrypted.seekg(0);
+    EXPECT_EQ(ctx_->CalculateChecksum(input_second), ctx_->CalculateChecksum(decrypted))
```
Можно вызывать CalculateChecksum для input, запомнить результат и тут его сравнить после всех операций. Тогда совсем аккуратно получится.

Nadezhda94 commented on tests/crypo_guard_ctx_test.cpp:
```diff
@@ -1,3 +1,242 @@
+        })
+        << "Тип исключения при дешифровке с неправильным ключом должен быть std::runtime_error";
+
+    // Дешифруем с правильным паролем
```
Тест хороший, проверяет нетривиальный сценарий, даже можно разделить этот тест на два: дешифровка с правильным и неправильным паролем.

ZernovNikolay replied:
Здесь суть теста заключается в проверке того, что в случае попытки дешифровки с неправильным паролем данные не инвалидируются. И в случае последующей попытки с верным паролем мы получим нужный результат

Nadezhda94 commented on tests/crypo_guard_ctx_test.cpp:
```diff
@@ -1,3 +1,242 @@
+        << "Дешифровка с неверным паролем не должна влиять на процесс работы с правильным паролем";
+}
+
+TEST_F(CipherTest, LargeDataEncryption) {
```
👍

Nadezhda94 commented on tests/crypo_guard_ctx_test.cpp:
```diff
@@ -1,3 +1,242 @@
+}
+
+TEST_F(CipherTest, InvalidInputOutputStreams) {
+    // 1. Невалидный входной поток (badbit)
```
Тоже можно разбить тест на несколько. Тем более что он как раз логически бьется на отдельные сценарии. Так проще отлаживать и поддерживать.

ZernovNikolay committed:
Version 2.0 of first sprint. Refactoring and fixing mistakes

ZernovNikolay commented:
Исправлены все ошибки и недочеты

ZernovNikolay committed:
Version 2.1 of first sprint.
Add [[nodiscard]] to Checksum methods
Add CheckAPIStability test for verify API

ZernovNikolay commented:
Добавлены улучшения, указанные в чате проекта на Практикуме
